source/agora/common/TransactionPool.d

```diff
@@ -0,0 +1,248 @@
+
+import d2sqlite3.database;
+import d2sqlite3.library;
+import d2sqlite3.results;
+import d2sqlite3.sqlite3;
+
+import vibe.core.log;
+
+import std.algorithm;
+import std.conv : to;
+import std.exception : collectException, enforce;
+import std.file : exists;
+import std.range;
+
+/// Ditto
+public class TransactionPool
+{
+    // should only be initialized once per process
+    shared static this()
+    {
+        .shutdown();
+        .config(SQLITE_CONFIG_MULTITHREAD);
+        version (unittest) {} else
+            .config(SQLITE_CONFIG_LOG, &loggerCallback, null);
+        .initialize();
+    }
+
+    /// SQLite db instance
+    private Database db;
+
+
+    /***************************************************************************
+
+        Params:
+            db_path = path to the database file, or in-memory storage if
+                      :memory: was passed
+
+    ***************************************************************************/
+
+    public this (in string db_path)
+    {
+        const db_exists = db_path.exists;
+        if (db_exists)
+            logInfo("Loading database from: %s", db_path);
+
+        // note: can fail. we may want to just recover txes from the network instead.
+        this.db = Database(db_path);
+
+        if (db_exists)
+            logInfo("Loaded database from: %s", db_path);
+
+        // create the table if it doesn't exist yet
+        this.db.execute("CREATE TABLE IF NOT EXISTS tx_pool " ~
+            "(key BLOB PRIMARY KEY, val BLOB NOT NULL)");
+    }
+
+    /***************************************************************************
+
+        Callback used
+
+        Params:
+            arg = unused (null)
+            code = the error code
+            msg = the error message
+
+    ***************************************************************************/
+
+    private static extern(C) void loggerCallback (void *arg, int code,
+        const(char)* msg) nothrow
+    {
+        try
+        {
+            logError("SQLite error: (%s) %s", code, msg.to!string);
+        }
+        catch (Exception ex)
+        {
+            // should not propagate exceptions to C code
+        }
+    }
+
+    /***************************************************************************
+
+        Shut down the database
+
+        Note: this method must be called explicitly, and not inside of
+        a destructor.
+
+    ***************************************************************************/
+
+    public void shutdown ()
+    {
+        this.db.close();
+    }
+
+    /***************************************************************************
+
+        Add a transaction to the pool
+
+        Params:
+            tx = the transaction to add
+
+        Throws:
+            SqliteException if the transaction failed to be added
+
+    ***************************************************************************/
+
+    public void add (Transaction tx) @trusted
```

Comment:
No `@trusted` at top level please

Reply:
What do you mean "top level" though?

Reply:
I mean, at the function level (unless the function is trivial).
Here you have no idea if `serializeFull` is safe for example.

Reply:
Right, so trusted delegates it is then. :)